### PR TITLE
Create polkamusic_milestone_1.md

### PR DESCRIPTION
**Pull request by cryptoshenoy**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [ ] The [invoice form :pencil:](https://forms.gle/vgQEZzvvFUjCyBiW6) has been filled out for this milestone. 
- [ ] This pull request is being made by the same account as the accepted application. 
- [ ] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application. 
- [ ] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/General-Grants-Program/blob/master/grants/milestone-deliverables-guidelines.md). 

Link to the application PR: https://github.com/w3f/Open-Grants-Program/pull/XXX < please fill this in. This is so that GitHub links to your delivery from the application.

### CODE REVIEW

**Comment by alxs** [comment]
Thank you for the delivery @cryptoshenoy. We will look into it as soon as possible.

**Comment by mmagician** [comment]
@cryptoshenoy After an initial round of testing, I found the following issues:
1. Multiple code warnings produced from `cargo clippy`, please fix.
2. Formatting is off and makes it harder to read, especially considering the amount of checks you do. Please keep a consistent style (e.g. try [rustfmt](https://github.com/rust-lang/rustfmt if you have no other personal preference).
2. Code duplication across your entire pallet. The following structure seems to make up for majority of the checks:
- ensure within range
- parse json & get value

```
ensure!(othercontractssharevalue <= 100, Error::<T>::InvalidOtherContractsShare); 	//check Composition Shares that must be <=100
// check other contracts quorum
let jsoq=crmdata.clone();
let othercontractsquorum=json_get_value(jsoq,"othercontractsquorum".as_bytes().to_vec());
let othercontractsquorum_slice=othercontractsquorum.as_slice();
let othercontractsquorum_str=match str::from_utf8(&othercontractsquorum_slice){
    Ok(f) => f,
    Err(_) => "100"
};
let othercontractsquorumvalue:u32 = match u32::from_str(othercontractsquorum_str){
    Ok(f) => f,
    Err(_) => 100,
};
```

Not only can the above be simplified (why are you first extracting slice, then string and then u32?), but you could probably come up with a succinct macro that would make your code more readable (it's very hard to follow your code in its present state). I would have tried to show you with an example, but there's no way for me to know whether the functionality wouldn't be altered as there's no unit tests to verify it, which brings me to the last point:
4. Unit and integration tests missing. With such a large amount of checks, it is essential to verify that your logic is correct.

**Comment by mmagician** [comment]
@cryptoshenoy Any updates here?

**Comment by cryptoshenoy** [comment]
Hey @mmagician. Will update you over here.

**Comment by cryptoshenoy** [comment]
Hello @mmagician . We have solved the issues and working on unit testing currently. Could you please have a look?

**Comment by mmagician** [comment]
@cryptoshenoy Thanks. As you see at the moment we have quite a lot of milestone deliveries ongoing (25 PRs open at the time of writing), so please allow an additional couple of days - I appreciate the patience.

**Comment by mmagician** [comment]
@cryptoshenoy I had a quick look and the code definitely looks much more readable. I will put this evaluation on hold for now, please let me know once you are ready with unit test!

**Comment by samuelelandi** [comment]
About point 1, cargo clippy is producing only 3 warnings now. I would not change the code for them, those suggestions does not seem correct.
Point 2, the formatting has been executed;
Point 3, I've created a function vecu8_to_u32() and replaced the ±10 lines with a single line calling it in all the possible points.
Point 4, testing unit. I've added it to the Pallet CRM (cargo test). There are tests for storing and reading contracts, storing and reading of change proposals and a subset of errors cases. I did not implement the testing for all the possible errors because they are very granular for easier debugger and may be excessive. If you want to see all the possible errors I will do. Please let me know.

**Comment by samuelelandi** [comment]
about this question "Not only can the above be simplified (why are you first extracting slice, then string and then u32?).
I did not find a better way to convert a number written in a Vec<u8> to u32 because of the current constraints with no-std libraries. Probably there are other ways as usual, by the way this one was looking reasonable to me. If you have different suggestion please let me know and I will apply to the code.

**Comment by mmagician** [comment]
@samuelelandi Thanks for the explanation, the changes you made look reasonable, at least at first glance. I will continue with the evaluation in the next couple of days.

**Comment by cryptoshenoy** [comment]
Hello @mmagician. The comments have been addressed and changes have been committed.

**Comment by mmagician** [comment]
Alright, I'll review it again soon. Thanks!

**Comment by cryptoshenoy** [comment]
Hello @mmagician. We will update the UX ASAP and update here.

**Comment by cryptoshenoy** [comment]
We have improved the UI significantly and addressed the issues. We are currently working on documentation and a video so that it's easier to understand.

**Comment by mmagician** [comment]
@cryptoshenoy Sounds good. Thanks for the update and let me know once you are ready for the next review.

**Comment by mmagician** [comment]
I just checked the newest portal, and it's nice that you added the option to choose Testnet vs. Local, thanks.
I will accept this milestone, since you have shown a lot of commitment indeed and in theory, have provided the instructions on how to use your application, so you fulfilled your obligations from the contract. 

That said, the UI is *still* very unintuitive, e.g. it's painful to input accounts in hex (why go for hex on polkadot in the first place??), and still no tiptools on the input fields. While the video might be instructive and useful for a grant evaluation, would you expect a new user to watch a 25 minute video just to use your platform? Imagine if you had to watch a 25 minute video to use Twitter or LinkedIn - the services would have died immediately. I would suggest getting a designer onboard who can really pivot your UX towards being user friendly, because at the moment it is quite the opposite. You had a nice idea with decentralised music management, but if you want your idea to take off, you should really revisit it from a non-technical & non-web3 user point of view. That's just my take on the matter, I hope some constructive criticism helps.
